README
Policy



INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Policy

0.001000 2015-05-06T12:00:16Z bd3c206
 - First version.



lib/Policy.pm
use 5.006;    # our
use strict;
use warnings;

package Policy;

# ABSTRACT: Policies, Standards and Guidelines for the world of Perl

our $VERSION = '0.001000';

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Policy - Policies, Standards and Guidelines for the world of Perl

=head1 VERSION

version 0.001000

=head1 ABOUT THIS NAMESPACE

C<Policy> is a namespace dedicated to documenting various current best practices and
recommendations for scopes as narrow as individual authors, or guidelines as broad as all of CPAN.

Existence of these policies are not necessarily gospel, but merely a canonicalization of human
knowledge in a more consumable form that doesn't require active discussion.

Members of a group may wish to enforce these standards within code they control, but there is
nothing stopping an individual consulting these standards, and adopting them as their own anyway.

In fact, encouraging the adoption of such standards, especially in regards to C<Toolchain> and
C<P5P> standards is highly encouraged.

=head1 SUB NAMESPACES

At this time, there are three primary categories:

=over 4

=item * L<< C<Policy::Org>|Policy::Org >> - Policies under the authority of specific organizations

Examples:

    Policy::Org::P5P
    Policy::Org::Toolchain
    Policy::Org::ShadowCat
    Policy::Org::Debian

B<NOTE:> There may be special cases for very high priority organizations within the CPAN ecosystem
such that the need for an C<Org::> prefix is optional. Both C<P5P> and C<Toolchain> are prime
candidates for a priority organization. However, the second of those two are unlikely to be classes
as high priority enough to warrant that naming.

=item * L<< C<Policy::Project>|Policy::Project >> - Policies dictated by specific projects

Examples:

    Policy::Project::Moose
    Policy::Project::DBIC

Projects typically can be distinguished from Organizations in that: Projects tend to be narrow
scoped to a very specific domain, whereas Organizations are broad having opinions that can apply to
almost anything.

Projects often gravitate around a particular namespace, such as the examples given, whereas
organizations are likely to be dabbling in a plethora of seemingly unrelated concerns.

=item * L<< C<Policy::Author>|Policy::Author >> - Policies dictated by specific authors

These entries should be strictly under the control of the named PAUSE identifier,
and people injecting under other peoples PAUSE identifier will have PAUSE admins gladly taking
control of things under that name that the individual did not authorize.

Examples:

    Policy::Author::AUTARCH
    Policy::Author::KENTNL
    Policy::Author::RIBASUSHI

=back

=head1 AUTHORSHIP OF NAMESPACES

The exact recommended layout of any policy sub-namespace, should in itself (at some stage) have a
policy encouraging a uniform layout. However, in the interim, my personal suggestion is as follows:

    Policy::<Category>::<Name>::<PolicyId>_<PolicyWordToken>

Where:

=over 4

=item * B<C<E<lt>CategoryE<gt>>> is one of C<Org|Project|Author>

=item * B<C<E<lt>NameE<gt>>> is your organizational identifier as discussed in
L<SUB NAMESPACES|/SUB NAMESPACES>

=item * B<C<E<lt>PolicyIdE<gt>>> is zero-padded 4-digit number such as C<0001>

=item * B<C<E<lt>PolicyWordTokenE<gt>>> is a simple word expression intended to give some sort of
identification, for example: C<SimpleVersions>

=back

The purpose here of the C<E<lt>PolicyIdE<gt>_E<lt>PolicyWordTokenE<gt>> is to enforce a simplified
visual and machine sort order that makes it clear the order in which the policies were added, to
make it obvious which policies were added more recently.

The namespace itself B<< C<< Policy::E<lt>CategoryE<gt>::E<lt>NameE<gt> >> >> should serve as an
index to the policies beneath it, grouping policies together by topic, severity, importance,
deprecation status, etc, each with a short blurb describing what will be found within.

Granted, such indexes are not entirely required, given the existence of MetaCPAN and friends,
which will generate a measure of index from package name and abstract alone. But it may be desired
to expand slightly beyond the terse requirements of abstracts, as well as being able to group
policies in multiple ways.

=head1 AUTHORSHIP OF POLICIES

Each policy is intended to be a "Living document", which should aim to remain "Current" at all
times.

Only if an entire document fails to make sense should the document itself be deprecated.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw(
   CPAN NAMESPACE namespace canonicalization NAMESPACES MetaCPAN
));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.036.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Policies, Standards and Guidelines for CPAN Modules",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Policy",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Policy",
  "PREREQ_PM" => {
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.89"
  },
  "VERSION" => "0.001000",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => "0.89",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
